MySql support.

SebastianStehle opened the pull request:
Closes  #6

SebastianStehle commented:
MySQL support.

I would like to reuse the test containers because it is really slow to restart them all the time. But the tests are not fully ready yet, but as preparation I created more random test data.

SebastianStehle commented:
I have just realized I am not done with the conflict handling yet.

PhenX commented on src/PhenX.EntityFrameworkCore.BulkInsert.MySql/PhenX.EntityFrameworkCore.BulkInsert.MySql.csproj:
```diff
@@ -0,0 +1,23 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <ItemGroup>
```
this item group is not needed

SebastianStehle replied:
Fixed

PhenX commented:
That's embarassing, it looks like the pipeline does not detect any test so it always tells that it's OK while it's not, with this pull request

SebastianStehle commented:
Benchmark added :)

SebastianStehle commented:
I hope I have fixed all tests.

There are few things worth to mention:

1. MySQL does not allow explicit conflicts, only general conflict statements. Therefore I tried to get a scenario running where no Match was defined. This is actually what I also need for upsert statements. Unfortunately SQL Server and Postgres do not allow it without a match, so I guess we have to find the primary keys for that.
2. MySQL has no support for "RETURNING" and no other methods to return the statements.
3. Autoincremented values are not added by default to the temp table, therefore you can also not find conflicts on that. I have added an option for that.
4. The tests were painfully slow and I could not run 10 sql servers with docker without docker causign problems. Therefore I have created a fixture for that.

PhenX commented:
Sure! I'll take a look asap

SebastianStehle commented:
I actually prefer the current solution with fixtures because it is super fast to run locally, but I think it is not related to the actual problem with the locking.

PhenX commented:
They are now actually IClassFixtures, it's still super fast to run locally, it seems to solve the locking issue two : too many containers where started at the same time

SebastianStehle commented:
The locking issue only happened with mysql and I think the problem was that the transaction was not disposed. At least it is working fine now. Everything is green and waiting for review.

PhenX commented on src/PhenX.EntityFrameworkCore.BulkInsert.MySql/MySqlBulkInsertProvider.cs:
```diff
@@ -0,0 +1,80 @@
+
+        var bulkCopy = new MySqlBulkCopy(connection, mySqlTransaction);
+        bulkCopy.DestinationTableName = tableName;
+        bulkCopy.BulkCopyTimeout = 60;
```
can you use the new option you added ?

SebastianStehle replied:
Done

PhenX commented on src/PhenX.EntityFrameworkCore.BulkInsert.MySql/MySqlDbContextOptionsExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+namespace PhenX.EntityFrameworkCore.BulkInsert.MySql;
+
+/// <summary>
+/// DbContext options extension for SQL Server.
```
wrong comment

SebastianStehle replied:
Solved

PhenX commented on src/PhenX.EntityFrameworkCore.BulkInsert.SqlServer/SqlServerBulkInsertProvider.cs:
```diff
@@ -22,7 +22,7 @@ public SqlServerBulkInsertProvider(ILogger<SqlServerBulkInsertProvider>? logger
     //language=sql
     /// <inheritdoc />
-    protected override string AddTableCopyBulkInsertId => $"ALTER TABLE {{0}} ADD {BulkInsertId} INT IDENTITY PRIMARY KEY;";
+    protected override string AddTableCopyBulkInsertId => $"ALTER TABLE {{0}} ADD {BulkInsertId} INT;";
```
I added this column as identity primary for future features like identity output, is it a problem for MySQL ? I'm not sure making it an option is a responsibility we want to expose to the user

SebastianStehle replied:
I have tested something and have forgotten to revert it. Good catch.

PhenX commented on tests/PhenX.EntityFrameworkCore.BulkInsert.Tests/DbContainer/TestDbContainerMySql.cs:
```diff
@@ -0,0 +1,35 @@
+    protected override IDatabaseContainer? GetDbContainer()
+    {
+        return new MySqlBuilder()
+            .WithReuse(true)
```
I'm not sure reuse works now, because of .WithName(GetRandomContainerName())

PhenX commented on tests/PhenX.EntityFrameworkCore.BulkInsert.Tests/DbContainer/TestDbContainer.cs:
```diff
@@ -39,29 +38,42 @@ public async Task InitializeAsync()
     }
 
-    public async Task DisposeAsync()
+    public async Task<TDbContext> CreateContextAsync()
```
Are all your changes required for mysql provider to work ?

SebastianStehle replied:
Not really, I added a few more tests, but most of the stuff is needed because of the "reuse" to isolate the tests logically from each other. It could be also solved by creating a new database with each run, but I like this, because it is easier to debug. You can just restart the container and connect with pgadmin or other tools.

PhenX replied:
Ok fair enough

SebastianStehle commented:
Nice, thank you very much for merging the branch and for your good work :)

PhenX commented:
@SebastianStehle and thank you for your contribution :)
